ext/pgsql/pgsql.c

```diff
@@ -3211,18 +3289,19 @@ PHP_FUNCTION(pg_copy_from)
 					if (UNEXPECTED(!tmp)) {
 						return;
 					}
-					query = (char *)emalloc(ZSTR_LEN(tmp) + 2);
-					strlcpy(query, ZSTR_VAL(tmp), ZSTR_LEN(tmp) + 2);
-					if (ZSTR_LEN(tmp) > 0 && *(query + ZSTR_LEN(tmp) - 1) != '\n') {
-						strlcat(query, "\n", ZSTR_LEN(tmp) + 2);
+					zend_string *zquery = zend_string_alloc(ZSTR_LEN(tmp) + 2, false);
+					memcpy(ZSTR_VAL(zquery), ZSTR_VAL(tmp), ZSTR_LEN(tmp) + 1);
+					ZSTR_VAL(zquery)[ZSTR_LEN(tmp) + 1] = 0;
+					if (ZSTR_LEN(zquery) > 0 && ZSTR_VAL(zquery)[ZSTR_LEN(tmp)]  != '\n') {
+						ZSTR_VAL(zquery)[ZSTR_LEN(tmp) ] = '\n';
 					}
-					if (PQputCopyData(pgsql, query, (int)strlen(query)) != 1) {
-						efree(query);
+					if (PQputCopyData(pgsql, ZSTR_VAL(tmp), ZSTR_LEN(tmp)) != 1) {
```

Comment:
The use of ZSTR_LEN(tmp) is wrong, because you need to include the '\n' if you added it in the if-check above.